### ISSUE TITLE
b3166db  Breaks seekRange()

### ISSUE THREAD

**Issue by vlinder** [shaka-project/shaka-player#3577]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes

**What version of Shaka Player are you using?**
3.1.1 => 3.2.0

**Can you reproduce the issue with our latest release version?**
Yes 3.2.0

**Can you reproduce the issue with the latest code from `master`?**
Yes.

**Are you using the demo app or your own custom app?**
Custom

**If custom app, can you reproduce the issue using our demo app?**
Unknown

**What browser and OS are you using?**
MacOS

**For embedded devices (smart TVs, etc.), what model and firmware version are you using?**
N/A

**What are the manifest and license server URIs?**
Not able to send a manifest url since they do timeout after 5 min. If you need anything else just throw me a comment and we might be able to work something out..

_(Edit: For the sake of GitHub issues performance, removed 100+kB of inline XML and attached here: 
[3577.mpd.txt](https://github.com/google/shaka-player/files/7197815/3577.mpd.txt) ~JCP)_

**What configuration are you using?  What is the output of `player.getConfiguration()`?**
```json
{
    "drm": {
        "retryParameters": {
            "maxAttempts": 4,
            "baseDelay": 1000,
            "backoffFactor": 2,
            "fuzzFactor": 1,
            "timeout": 30000,
            "stallTimeout": 5000,
            "connectionTimeout": 10000
        },
        "servers": {
            "com.widevine.alpha": "[redacted]"
        },
        "clearKeys": {},
        "advanced": {},
        "delayLicenseRequestUntilPlayed": false,
        "logLicenseExchange": false,
        "updateExpirationTime": 1,
        "preferredKeySystems": []
    },
    "manifest": {
        "retryParameters": {
            "maxAttempts": 2,
            "baseDelay": 1000,
            "backoffFactor": 2,
            "fuzzFactor": 0.5,
            "timeout": 30000,
            "stallTimeout": 5000,
            "connectionTimeout": 10000
        },
        "availabilityWindowOverride": null,
        "disableAudio": false,
        "disableVideo": false,
        "disableText": false,
        "disableThumbnails": false,
        "defaultPresentationDelay": 0,
        "dash": {
            "clockSyncUri": "https://time.akamai.com/?iso",
            "ignoreDrmInfo": false,
            "disableXlinkProcessing": false,
            "xlinkFailGracefully": false,
            "ignoreMinBufferTime": false,
            "autoCorrectDrift": true,
            "initialSegmentLimit": 1000,
            "ignoreSuggestedPresentationDelay": false,
            "ignoreEmptyAdaptationSet": false,
            "ignoreMaxSegmentDuration": false,
            "keySystemsByURI": {
                "urn:uuid:1077efec-c0b2-4d02-ace3-3c1e52e2fb4b": "org.w3.clearkey",
                "urn:uuid:edef8ba9-79d6-4ace-a3c8-27dcd51d21ed": "com.widevine.alpha",
                "urn:uuid:9a04f079-9840-4286-ab92-e65be0885f95": "com.microsoft.playready",
                "urn:uuid:79f0049a-4098-8642-ab92-e65be0885f95": "com.microsoft.playready",
                "urn:uuid:f239e769-efa3-4850-9c16-a903c6932efb": "com.adobe.primetime"
            }
        },
        "hls": {
            "ignoreTextStreamFailures": false,
            "ignoreImageStreamFailures": false,
            "useFullSegmentsForStartTime": false
        }
    },
    "streaming": {
        "retryParameters": {
            "maxAttempts": 2,
            "baseDelay": 1000,
            "backoffFactor": 2,
            "fuzzFactor": 0.5,
            "timeout": 30000,
            "stallTimeout": 5000,
            "connectionTimeout": 10000
        },
        "rebufferingGoal": 2,
        "bufferingGoal": 10,
        "bufferBehind": 30,
        "ignoreTextStreamFailures": false,
        "alwaysStreamText": true,
        "startAtSegmentBoundary": false,
        "gapDetectionThreshold": 0.1,
        "smallGapLimit": 0.5,
        "jumpLargeGaps": true,
        "durationBackoff": 1,
        "forceTransmuxTS": false,
        "safeSeekOffset": 5,
        "stallEnabled": true,
        "stallThreshold": 1,
        "stallSkip": 0.1,
        "useNativeHlsOnSafari": true,
        "inaccurateManifestTolerance": 2,
        "lowLatencyMode": false,
        "autoLowLatencyMode": false,
        "forceHTTPS": false,
        "preferNativeHls": false,
        "updateIntervalSeconds": 1
    },
    "offline": {
        "usePersistentLicense": true
    },
    "abr": {
        "enabled": true,
        "useNetworkInformation": true,
        "defaultBandwidthEstimate": 5000000,
        "switchInterval": 4,
        "bandwidthUpgradeTarget": 0.85,
        "bandwidthDowngradeTarget": 0.95,
        "restrictions": {
            "minWidth": 0,
            "maxWidth": null,
            "minHeight": 0,
            "maxHeight": null,
            "minPixels": 0,
            "maxPixels": null,
            "minFrameRate": 0,
            "maxFrameRate": null,
            "minBandwidth": 0,
            "maxBandwidth": null
        }
    },
    "preferredAudioLanguage": "",
    "preferredTextLanguage": "",
    "preferredVariantRole": "",
    "preferredTextRole": "",
    "preferredAudioChannelCount": 2,
    "preferredVideoCodecs": [],
    "preferredAudioCodecs": [],
    "preferForcedSubs": false,
    "preferredDecodingAttributes": [],
    "restrictions": {
        "minWidth": 0,
        "maxWidth": null,
        "minHeight": 0,
        "maxHeight": null,
        "minPixels": 0,
        "maxPixels": null,
        "minFrameRate": 0,
        "maxFrameRate": null,
        "minBandwidth": 0,
        "maxBandwidth": null
    },
    "playRangeStart": 0,
    "playRangeEnd": null
}
```

**What did you do?**
We calculate the duration from `seekRange()` `end - start` for all our streams, including VOD. This now has a big mismatch against `video.duration` for our VOD manifests (have not tried live). We have DAI being inserted but this has not been an issue with this previously.

We did a git bisect and this commit introduces the issue in master: https://github.com/google/shaka-player/commit/b3166db4bc7cac968a03796bc9c2d7e598fb88c7

**What did you expect to happen?**
Duration calculation from seekRange to be practically the same as `mediaPresentationDuration` in the MPD.

**What actually happened?**
MPD duration is around 2760 seconds. `seekRange()` `end - start` is around 2300 seconds. `start` is `0` in this case.

Did a manual summation of all the Periods in the manifest. The pretty much sum up to the expected duration.

**Comment by theodab** [comment]
The change that you highlighted causes us to prefer to go with the end time of the last segment, if available, rather than the duration provided in the manifest. If the output of `seekRange` has become significantly shorter, I suppose that means that some of the segments might be not being counted?

Just to be sure, if you access this variable (while loading the player in debug mode):
```js
player.getManifest().presentationTimeline.maxSegmentEndTime_
```
Does it report 2300 seconds?

**Comment by martinstark** [comment]
Today's findings:

Me and a colleague both started playing a dynamic manifest at about the same time.

Person 1 start time: Fri, 21 Jan 2022 09:54:28 GMT
Person 2 start time: Fri, 21 Jan 2022 09:53:02 GMT

We started the stream roughly 86 seconds apart.

When the manifest turned static we both got identical manifests, shaka's `maxSegmentEndTime`, `duration` and `seekRange().end` all show `3601.877355`. Both of our streams stalled, but at different times:

Person 1 stall time: 3566.986595
Person 2 stall time: 3482.72427

The streams stalled roughly 84.2 seconds apart. Accounting for differing start up times this is suspiciously close to 86.

For a later attempt on a different stream, my colleague got a stall and I didn't. With the same manifest, loaded at roughly the same time.

I get a feeling we're running into a race condition, which may or may not be related to the `maxSegmentEndTime_` problem we saw on an earlier shaka version. The root cause still appears to be changes made in shaka 3.1.2.

Since the symptoms now looks unrelated to the exact issue in this particular thread I'll look creating a different issue report once we can dig deeper.
